Fixes to CMS failures in Django 1.11 tests

bmedx opened the pull request:
Should fix remaining CMS test failures in 1.11, also includes a refactor of student views.

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

michaelyoungstrom commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

michaelyoungstrom commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented on cms/djangoapps/contentstore/tasks.py:
```diff
@@ -270,8 +270,15 @@ def create_export_tarball(course_module, course_key, context, status=None):
 
     Updates the context with any error information if applicable.
     """
+    export_path = settings.MEDIA_ROOT / 'exports'
```
This is potentially problematic and worth discussion. Currently we are creating these temporary export files in /tmp then using Django Storages to move them to the configured semi-permanent download location (S3 for edx.org). Django 1.11 has enhanced security to prevent it from allowing FileFields to access files from anywhere outside of settings.MEDIA_ROOT, which is the publicly accessible upload path. As such this change makes these export files available in that directory path during creation (albeit in randomly named files). Mitigation options for this off-hand include:

- Locking down security on the MEDIA_ROOT / exports url, which is likely to be error prone and forgotten by partners.
- Writing a custom FileField implementation that lets us override the security feature to use a different path.

jmbowman replied:
Where exactly is it hitting the problem?  From my understanding, the current workflow should have been something like this:

1. Export tarball is created using generic Python code under /tmp
1. Django is told the filename and given the content of the file
1. Django copies it to an appropriate location for the current storage system (which is under MEDIA_ROOT for the default one) when saving the model

bmedx replied:
It happens in [step 2](https://github.com/edx/edx-platform/blob/113abb5bff1bb0299adbc505c67c13802f0a9495/cms/djangoapps/contentstore/tasks.py#L257), when the contents of the file are read into a Django File for the UserTaskArtifact.

jmbowman replied:
Looking at one of the failures for that, I wonder if we're sending it "/tmp/filename.tar.gz" as the filename, which is tripping up the join operation and just ignoring the intended path prefix.  Could you try extracting just the filename from "tarball.name" on that line, and pass that in as the first argument to `save()`?

bmedx replied:
Seems to have worked, thanks!

bmedx commented:
jenkins run django upgrade python

jmbowman commented on cms/djangoapps/contentstore/tests/test_contentstore.py:
```diff
@@ -1442,7 +1445,7 @@ def test_create_course_with_course_creator(self):
             self.assert_created_course()
 
-    def test_create_course_with_unicode_in_id_disabled(self):
+    def test_create_course_with_text_type_in_id_disabled(self):
```
Given that the `ALLOW_UNICODE_COURSE_ID` setting name didn't change, probably should leave the test name as it was.

bmedx replied:
Bah, bad replace. Will fix.

jmbowman commented on cms/djangoapps/contentstore/views/tests/test_library.py:
```diff
@@ -324,7 +324,7 @@ def test_manage_library_users(self):
         self.assertEqual(response.status_code, 200)
         # extra_user has not been assigned to the library so should not show up in the list:
-        self.assertNotIn(extra_user.username, response.content)
+        self.assertNotIn(str(extra_user.username), response.content)
```
I don't think this will work for Python 3; maybe use `six.binary_type(extra_user.username)` instead?

bmedx replied:
Cleaned up that and the uses of unicode

jmbowman commented on openedx/core/djangolib/rate_limit_all_user_auth_backend.py:
```diff
@@ -0,0 +1,20 @@
+# TODO: Remove Django 1.11 upgrade shim
+# SHIM: Post-upgrade we should only need the AllowAllUsersModelBackend.
+try:
+    from django.contrib.auth.backends import AllowAllUsersModelBackend as django_backend
```
Where is this needed?  We introduced `NEW_USER_AUTH_BACKEND` in `common/djangoapps/student/views.py` specifically to avoid needing to allow inactive users to log in except when they're first created.

bmedx replied:
Ah interesting, I did think we'd addressed it but I didn't see that backend. The errors I was addressing were `common.djangoapps.student.tests.test_login/LoginTest/test_login_not_activated` and `not_activated_no_pii`. They were failing on bad username / pass authenticating. I'll see if I can refactor to use that.

doctoryes commented on cms/djangoapps/contentstore/management/commands/delete_course.py:
```diff
@@ -57,7 +57,11 @@ def add_arguments(self, parser):
             # a course key may have unicode chars in it
-            course_key = text_type(options['course_key'], 'utf8')
+            try:
+                course_key = text_type(options['course_key'], 'utf8')
```
What does this decoding support? The possible sending of encoded Unicode chars into course_key? On my LMS shard 4 branch, I fixed this same error by just dropping the `, 'utf8'`. This change is safer but is it necessary?

bmedx replied:
Seems like it was added intentionally to cover that case: https://github.com/edx/edx-platform/commit/8bc7ae146690d05321cdeeaef7694afd85789d55#diff-6e2de8bda28b18a6d50bc686d199e81bL53

doctoryes commented on cms/djangoapps/course_creators/tests/test_admin.py:
```diff
@@ -105,7 +106,13 @@ def check_admin_message_state(state, expect_sent_to_admin, expect_sent_to_user):
-                context = {'user_name': "test_user", 'user_email': u'test_user+courses@edx.org'}
+                # TODO: Remove Django 1.11 upgrade shim
+                # SHIM: Usernames come back as unicode in 1.10+, remove this shim post-upgrade
+                if django.VERSION[0] == 1 and django.VERSION[1] < 10:
```
Nit: I'd prefer `if django.VERSION < (1, 10):`

bmedx replied:
Doh.

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
@jmbowman @doctoryes This should be ready for re-review now. I ended up needing to do some shimming in login that could particularly use some attention.

jmbowman commented on common/djangoapps/student/views.py:
```diff
@@ -1564,6 +1564,32 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
+        # and the duplicated code at the end of this function.
+        #
+        # Note that this will be triggered for ALL inactive users, regardless of whether the password is correct. In
+        # either case they need to activate before continuing so password status is moot.
```
My brain is starting to hurt trying to trace through all this.  The `login_user` method is now 251 lines long; can the duplicated portion be broken out into a separate function, that's just called in 2 different places?  Feel free to break out other blocks while you're at it...

jmbowman commented on lms/envs/common.py:
```diff
@@ -661,7 +661,7 @@ def _add_microsite_dirs_to_default_template_engine(settings):
 
 # use the ratelimit backend to prevent brute force attacks
-AUTHENTICATION_BACKENDS = ['ratelimitbackend.backends.RateLimitModelBackend']
+AUTHENTICATION_BACKENDS = ['ratelimitbackend.backends.RateLimitModelBackend', ]
```
This isn't really hurting anything, but can this just be reverted so there are no changes for this file?

doctoryes commented on common/djangoapps/student/views.py:
```diff
@@ -1564,6 +1564,32 @@ def login_user(request, error=""):  # pylint: disable=too-many-statements,unused
+            return JsonResponse({
+                "success": False,
+                "value": _generate_not_activated_message(user_found_by_email_lookup),
+            })  # TODO: this should be status code 400  # pylint: disable=fixme
```
Dang it. I just spent some time on a solution for this issue as well:
https://github.com/edx/edx-platform/pull/17130/commits/47c48c6b261f2d443cd2485673b4fd62a17fa7c2
But this solution seems more targeted and therefore better.

doctoryes replied:
I will now check this PR before attempting any further LMS shard 4 fixes - lesson learned!

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
@jmbowman @doctoryes @pwnage101 Given the knowledge that I'll have to back-merge some changes coming down today, this is probably good for re-review. Especially relevant is the refactor of login/logout into the new views_login.py.

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run all

jmbowman commented on pavelib/paver_tests/test_stylelint.py:
```diff
@@ -13,11 +13,6 @@ class TestPaverStylelint(PaverTestCase):
-
-    def setUp(self):
-        super(TestPaverStylelint, self).setUp()
-        pass
```
Sigh...this is the kind of thing that makes you want to look in the history and see just how it ended up there in the first place.

bmedx commented:
jenkins run django upgrade python

bmedx commented:
jenkins run django upgrade python

bmedx commented on common/djangoapps/student/views.py:
```diff
@@ -2833,51 +2256,6 @@ def password_reset_confirm_wrapper(request, uidb36=None, token=None):
-        return JsonResponse({
-            "success": False,
-            "error": _('No inactive user with this e-mail exists'),
-        })  # TODO: this should be status code 400  # pylint: disable=fixme
```
I pulled them in the move as they were over 4 years old. If at some point we decide to do them the comment up top says where to find older history, but I think this whole section is being deprecated anyway.

bmedx commented:
jenkins run django upgrade python

bmedx commented:
Ok @jmbowman I refactored the student views into a new directory and fixed things up. I'm re-running the 1.11 tests because LMS 1 failed on the warning output, but I don't think there were any failures there. Want to take a look again?

bmedx commented:
jenkins run django upgrade python

jmbowman commented on common/djangoapps/student/tests/test_login.py:
```diff
@@ -499,16 +537,16 @@ def test_externalauth_login_required_course_context(self):
 
-        TARGET_URL_SHIB = reverse('courseware', args=[text_type(self.shib_course.id)])  # pylint: disable=invalid-name
-        shib_response = self.client.get(**{'path': TARGET_URL_SHIB,
+        target_url_shib = reverse('courseware', args=[text_type(self.shib_course.id)])  # pylint: disable=invalid-name
```
The pylint comment here should't be needed anymore.

jmbowman commented on common/djangoapps/student/urls.py:
```diff
@@ -6,50 +6,50 @@
 from django.contrib.auth.views import password_reset_complete
 
-import student.views
+import views
```
Would `from . import views` be better?  Python 3 no longer supports implicit relative imports.

bmedx commented:
jenkins run django upgrade python
